### ISSUE TITLE
#includes with "../" often point at non-existent file.

### ISSUE THREAD

**Issue by deeglaze** [openenclave/openenclave#2146]
I'm working on getting openenclave to build in Bazel, and I'm finding that it does not like up paths in cases where -I flags are needed to resolve file locations. There are > 100 such up file includes, and I haven't resolved them all yet, but I think it could be helpful for readability and Bazel support to resolve those paths in a repo-root-relative manner.

For example, in `enclave/sgx/qeidinfo.c` there is the line

```c
#include "../common/sgx/qeidentity.h"
```

which resolves naively and incorrectly to `enclave/common/sgx/qeidentity.h`. A repo-root-relative include path would be `common/sgx/qeidentity.h`. I presume this particular include resolves in your build due to a lurking `-Icommon` or `-Ihost` that would compose with `..` to find the file.

Bazel parses includes and sandboxes builds to only allow files that are explicitly stated as dependencies. It helps organizational hygiene, but it's definitely frustrating before you get used to it.

**Comment by deeglaze** [comment]
I see also that -I flag order will matter too, for files such as host/sgx/ocalls.c, which #includes two different `ocall.h` files. repo-root-relative path structure would help disambiguate these.

**Comment by andschwa** [comment]
Oh, interesting. In general I agree, I think repo-root-relative is often a cleaner approach. Let me tag @mikbras and @CodeMonkeyLeet in here.

**Comment by CodeMonkeyLeet** [comment]
I tend to agree that we should use repo-root-relative paths and specifically in this case we should not allow lateral relative traversals out of the current directory.

Ideally, local headers should be local headers and any shared headers should be organized under the broader `include` folder (so most references are relative to root of `include`).

There's a pattern of use for scoping headers relative to the file hierarchy (e.g. `host/sgx/calls.c` referencing `../calls.h` shared between SGX and OPTEE implementations.) I'm kind of okay with that, but the more uniform approach would be to roll them up under `include/openenclave/internal` ... honestly, we need to taxonomize headers in that folder anyway.

**Comment by deeglaze** [comment]
I'm finding myself taxonomizing the internal headers myself, since I end up pairing the source that defines symbols with headers that declare them. Another interesting thing about Bazel in that compilation defines have to be in the target definition itself, so headers y'all have that have two different interpretations depending on something like `defined(OE_BUILD_ENCLAVE)` have to be built in two different targets. The dependencies permeate through, too, so you end up with targets that are

*   truly common (e.g., common/result.c and openenclave/internal/result.h) and can be used regardless of host or enclave modality, or
*   host/enclave-dependent by design (the headers should only be interpretable by an enclave or the host, e.g., tee_u.c and tee_u.h), or
*   host/enclave-dependent by happenstance (the source or headers expand differently based on C-preprocessor defines, e.g., common/argv.c and openenclave/internal/argv.h).
*   host/enclave-dependent by symbol definition selection (e.g., tracee.c/traceh.c and openenclave/internal/trace.h)

The latter two are similar, but I find the third (happenstance) to be the most problematic for Bazel builds. Typically we want to keep files meaning the same thing throughout a project. The last bullet can be handled with build configuration that selects implementation files based on the platform.

If a target depends on a dependent target, then it becomes dependent, and the target definitions combinatorially expand.

**Comment by CodeMonkeyLeet** [comment]
I think the preference being expressed is that headers and source files should be decomposed into distinct units of functionality and then assembled by the build definitions without reliance on preprocessor definitions, is that accurate? That seems like a distinct issue from the organization of headers in the file hierarchy and we might want to spin up a separate issue for that. 

A couple of questions to clarify my understanding of the problem and what the desired solution might be:

1. I was thinking of the taxonomy in terms of the range of dimensions that also depend on preprocessor definitions, including target architecture, target OS, & compiler toolchain. How does Bazel deal with these, which are fairly common in headers & source files outside of the OE SDK? I'd like to get a sense for how unique the pain points are relative to the the more general examples of doing make/ninja to Bazel translations.

2. I'm not sure I understand the "host/enclave-dependent by happenstance case" ... isn't that any use of preprocessor directives as set by the build environment?
    1. Is the only difference between bullets 3 & 4 that 3 covers redefinition of symbols based on preprocessor directives whereas 4 is used to inject additional code?

**Comment by deeglaze** [comment]
Ah, I should note that Bazel does not allow dependents to alter the build environment. A host target that depends on "common" will get the same library bits that an enclave that defines "OE_BUILD_ENCLAVE" would get. Compiler options are local to a target. The "common" sources have to be compiled in a separate target that explicitly names the compiler options that it needs, like "-DOE_BUILD_ENCLAVE=1".

A target means one thing based on top level build flags. There is no way for a target to have multiple interpretations (host/enclave) without experimental features (e.g., transitions) that allow changing overall configuration in largely disruptive ways.

The difference between 3 & 4 is specifically separating multimodal source and multiplatform source. There can be multiple modes on a single platform (host/enclave), but only one platform selection at a time. CPU or TEE backend is one aspect of a platform constraint, but so is something like USE_LIBSGX, or in our case, USE_BORINGSSL, since we'll be writing an enclave/crypto_bssl alternative to y'all's mbedtls use in enclave/crypto.

One pattern that we've found helpful for "public" versus "private" headers, and avoiding deep tree references in users' #include directives is to have the top level "include/" directory populated with "forwarding headers" that #include the appropriate header within the API source tree that it stands for and seeks to make public. API-internal code would not use the public headers, only examples of using the API.

For target architecture, OS and toolchain, we have the concept of a crosstool, which defines multiple platform-specific ways of invoking whichever selected compiler using its platform matching process. It's similar to CMake's pkg-config, but it's a more pervasive concept when in comes to third-party dependencies. Third-party dependencies have their own Bazel BUILD files that provide local compiler settings etc on each target, but when they're built with a particular crosstool, the crosstool default flags apply first. We tend not to have to modify our dependencies at all and just reuse their build definitions.

**Comment by CodeMonkeyLeet** [comment]
Thanks for the explanation, it's really helpful. The distinction between 3 & 4 being multimodal in the Bazel paradigm is interesting, because while we build both host & enclave for a Linux platform as a single action, for Windows builds, the enclave and host would be separate platforms (i.e. they build under separate tool chains for separate target OS, since we only compile enclaves as ELF today); I can see how that's not intuitive though.

Just to confirm, the proposed solution for the multimodal case is to separate out sources so that you can define only a single platform, but build host and enclave with the same platform configuration via source selection, correct?

**Comment by deeglaze** [comment]
Yes, that's correct. We're also planning on adding a new triple to clang in order to target sgx builds, since existing triples define the wrong processor symbols, and the unknown triple falls back on gcc to link. It makes sense to us in the Asylo team to consider enclaves to be a distinct platform from the host operating system.

**Comment by yakman2020** [comment]
Could we create specific issues for this with specific action items?

**Comment by CodeMonkeyLeet** [comment]
I'd break the `TODO` list in order of priority:

- [ ] Fix existing `#include` statements that have incorrect relative paths
- [ ] Refactor shared headers that are referenced using relative paths into root relative path `include/openenclave/internal`
- [ ] Organize the set of `include/openenclave/internal` headers (e.g. as has been done with `include/openenclave/internal/crypto`
    - @mikbras Feel free to spin this out into a separate issue
- [ ] Refactor host/enclave source files into distinct files instead of relying on `OE_BUILD_ENCLAVE` definitions.
    - @deeglaze This is a little more invasive and for the moment I'd recommend just treating host/enclave as separate platforms.
